File: src/main/java/ltdjms/discord/gametoken/services/DiceGame2Service.java
package ltdjms.discord.gametoken.services;

import ltdjms.discord.currency.domain.CurrencyTransaction;
import ltdjms.discord.currency.domain.MemberCurrencyAccount;
import ltdjms.discord.currency.persistence.MemberCurrencyAccountRepository;
import ltdjms.discord.currency.services.CurrencyTransactionService;
import ltdjms.discord.gametoken.domain.DiceGame2Config;
import ltdjms.discord.shared.events.BalanceChangedEvent;
import ltdjms.discord.shared.events.DomainEventPublisher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Service for the dice-game-2 mini-game.
 * Handles dice rolling, reward calculation with straights and triples, and currency distribution.
 *
 * <p>The number of dice rolled is determined by the number of tokens spent:
 * 1 token = 3 dice.</p>
 *
 * <p>Reward rules:</p>
 * <ul>
 *   <li>Straights (consecutive increasing sequence of length >= 3): sum × straightMultiplier</li>
 *   <li>Triples (exactly 3 consecutive same values):
 *     <ul>
 *       <li>Sum < 10 (values 1-3): tripleLowBonus</li>
 *       <li>Sum >= 10 (values 4-6): tripleHighBonus</li>
 *     </ul>
 *   </li>
 *   <li>4+ consecutive same values: NOT a triple, counted as non-straight</li>
 *   <li>Non-straight/non-triple dice: sum × baseMultiplier</li>
 * </ul>
 */
public class DiceGame2Service {

    private static final Logger LOG = LoggerFactory.getLogger(DiceGame2Service.class);

    private final MemberCurrencyAccountRepository currencyRepository;
    private final CurrencyTransactionService transactionService;
    private final DomainEventPublisher eventPublisher;
    private final Random random;

    public DiceGame2Service(
            MemberCurrencyAccountRepository currencyRepository,
            CurrencyTransactionService transactionService,
            DomainEventPublisher eventPublisher) {
        this(currencyRepository, transactionService, eventPublisher, new Random());
    }

    /**
     * Constructor with injectable Random for testing.
     */
    public DiceGame2Service(
            MemberCurrencyAccountRepository currencyRepository,
            CurrencyTransactionService transactionService,
            DomainEventPublisher eventPublisher,
            Random random) {
        this.currencyRepository = currencyRepository;
        this.transactionService = transactionService;
        this.eventPublisher = eventPublisher;
        this.random = random;
    }

    /**
     * Plays the dice game for a member using the provided configuration and token amount.
     * Rolls dice (3 per token) and calculates the total reward based on straights, triples, and remaining dice.
     * The reward is added to the member's currency account.
     *
     * @param guildId   the Discord guild ID
     * @param userId    the Discord user ID
     * @param config    the game configuration containing multipliers and bonuses
     * @param diceCount the number of dice to roll (equals tokensSpent * 3)
     * @return the game result
     */
    public DiceGame2Result play(long guildId, long userId, DiceGame2Config config, int diceCount) {
        LOG.debug("Playing dice-game-2 for guildId={}, userId={}, diceCount={}, straightMult={}, baseMult={}, tripleLow={}, tripleHigh={}",
                guildId, userId, diceCount, config.straightMultiplier(), config.baseMultiplier(),
                config.tripleLowBonus(), config.tripleHighBonus());

        // Roll dice
        List<Integer> diceRolls = rollDice(diceCount);

        // Analyze the rolls with configured multipliers
        RewardAnalysis analysis = analyzeRolls(diceRolls, config);

        // Calculate total reward
        long totalReward = analysis.totalReward();

        // Apply reward to currency account
        long previousBalance = currencyRepository.findOrCreate(guildId, userId).balance();
        applyRewardToCurrency(guildId, userId, totalReward);
        long newBalance = currencyRepository.findByGuildIdAndUserId(guildId, userId)
                .map(MemberCurrencyAccount::balance)
                .orElse(previousBalance + totalReward);

        // Record transaction
        if (totalReward > 0) {
            transactionService.recordTransaction(
                    guildId,
                    userId,
                    totalReward,
                    newBalance,
                    CurrencyTransaction.Source.DICE_GAME_2_WIN,
                    null
            );
        }
        
        // Publish event
        eventPublisher.publish(new BalanceChangedEvent(guildId, userId, newBalance));

        DiceGame2Result result = new DiceGame2Result(
                guildId,
                userId,
                diceRolls,
                totalReward,
                previousBalance,
                newBalance,
                analysis.straightSegments(),
                analysis.tripleSegments(),
                analysis.straightReward(),
                analysis.nonStraightReward(),
                analysis.tripleReward()
        );

        LOG.info("Dice game 2 completed: guildId={}, userId={}, rolls={}, reward={}, newBalance={}",
                guildId, userId, diceRolls, totalReward, newBalance);

        return result;
    }

    /**
     * Rolls the specified number of dice for a game.
     *
     * @param count the number of dice to roll
     * @return list of dice values (1-6)
     */
    List<Integer> rollDice(int count) {
        List<Integer> rolls = new ArrayList<>(count);
        for (int i = 0; i < count; i++) {
            rolls.add(random.nextInt(6) + 1);  // 1-6
        }
        return rolls;
    }

    /**
     * Analyzes the dice rolls to identify straights, triples, and calculate rewards
     * using configured multipliers.
     *
     * @param diceRolls the list of dice values
     * @param config    the game configuration
     * @return the analysis result
     */
    RewardAnalysis analyzeRolls(List<Integer> diceRolls, DiceGame2Config config) {
        boolean[] usedInStraight = new boolean[diceRolls.size()];
        boolean[] usedInTriple = new boolean[diceRolls.size()];

        // First pass: identify straights (consecutive increasing sequences of length >= 3)
        List<List<Integer>> straightSegments = findStraights(diceRolls, usedInStraight);

        // Second pass: identify triples (exactly 3 consecutive same values)
        // Must not overlap with straights
        List<List<Integer>> tripleSegments = findTriples(diceRolls, usedInStraight, usedInTriple);

        // Calculate rewards using configured multipliers
        long straightReward = calculateStraightReward(straightSegments, config.straightMultiplier());
        long tripleReward = calculateTripleReward(tripleSegments, config.tripleLowBonus(), config.tripleHighBonus());
        long nonStraightSum = calculateNonStraightSum(diceRolls, usedInStraight, usedInTriple);
        long nonStraightReward = nonStraightSum * config.baseMultiplier();

        long totalReward = straightReward + nonStraightReward + tripleReward;

        return new RewardAnalysis(
                straightSegments,
                tripleSegments,
                straightReward,
                nonStraightReward,
                tripleReward,
                totalReward
        );
    }

    /**
     * Finds all straight segments (consecutive increasing sequences of length >= 3).
     */
    private List<List<Integer>> findStraights(List<Integer> diceRolls, boolean[] usedInStraight) {
        List<List<Integer>> straights = new ArrayList<>();

        int i = 0;
        while (i < diceRolls.size()) {
            int start = i;
            // Find the longest increasing sequence starting at i
            while (i + 1 < diceRolls.size() && diceRolls.get(i + 1) == diceRolls.get(i) + 1) {
                i++;
            }
            int length = i - start + 1;

            if (length >= 3) {
                List<Integer> segment = new ArrayList<>();
                for (int j = start; j <= i; j++) {
                    segment.add(diceRolls.get(j));
                    usedInStraight[j] = true;
                }
                straights.add(segment);
            }
            i++;
        }

        return straights;
    }

    /**
     * Finds all triple segments (exactly 3 consecutive same values).
     * A run of 4+ consecutive same values is NOT a triple.
     * Triples cannot overlap with straights.
     */
    private List<List<Integer>> findTriples(List<Integer> diceRolls, boolean[] usedInStraight, boolean[] usedInTriple) {
        List<List<Integer>> triples = new ArrayList<>();

        int i = 0;
        while (i < diceRolls.size()) {
            // Skip if already used in a straight
            if (usedInStraight[i]) {
                i++;
                continue;
            }

            int start = i;
            int value = diceRolls.get(i);

            // Count consecutive same values
            while (i + 1 < diceRolls.size() &&
                    diceRolls.get(i + 1).equals(value) &&
                    !usedInStraight[i + 1]) {
                i++;
            }

            int length = i - start + 1;

            // Exactly 3 consecutive same values = triple
            if (length == 3) {
                List<Integer> segment = new ArrayList<>();
                for (int j = start; j <= i; j++) {
                    segment.add(diceRolls.get(j));
                    usedInTriple[j] = true;
                }
                triples.add(segment);
            }
            // If length > 3, it's NOT a triple - these dice will be counted as non-straight

            i++;
        }

        return triples;
    }

    /**
     * Calculates the reward for straight segments using configured multiplier.
     */
    private long calculateStraightReward(List<List<Integer>> straightSegments, long straightMultiplier) {
        long sum = 0;
        for (int i = 0; i < straightSegments.size(); i++) {
            List<Integer> segment = straightSegments.get(i);
            for (int j = 0; j < segment.size(); j++) {
                sum += segment.get(j);
            }
        }
        return sum * straightMultiplier;
    }

    /**
     * Calculates the reward for triple segments using configured bonuses.
     */
    private long calculateTripleReward(List<List<Integer>> tripleSegments, long tripleLowBonus, long tripleHighBonus) {
        long reward = 0;
        for (int i = 0; i < tripleSegments.size(); i++) {
            List<Integer> segment = tripleSegments.get(i);
            // Each triple segment always has exactly 3 elements with the same value
            int sum = segment.get(0) * 3;
            if (sum < 10) {
                reward += tripleLowBonus;
            } else {
                reward += tripleHighBonus;
            }
        }
        return reward;
    }

    /**
     * Calculates the sum of dice values not used in straights or triples.
     */
    private long calculateNonStraightSum(List<Integer> diceRolls, boolean[] usedInStraight, boolean[] usedInTriple) {
        long sum = 0;
        for (int i = 0; i < diceRolls.size(); i++) {
            if (!usedInStraight[i] && !usedInTriple[i]) {
                sum += diceRolls.get(i);
            }
        }
        return sum;
    }

    /**
     * Applies the reward to the member's currency account.
     * If the reward exceeds the max adjustment amount, splits into multiple adjustments.
     */
    void applyRewardToCurrency(long guildId, long userId, long totalReward) {
        long remaining = totalReward;
        long maxAdjustment = MemberCurrencyAccount.MAX_ADJUSTMENT_AMOUNT;

        while (remaining > 0) {
            long adjustment = Math.min(remaining, maxAdjustment);
            currencyRepository.adjustBalance(guildId, userId, adjustment);
            remaining -= adjustment;
        }
    }

    /**
     * Internal record for reward analysis results.
     */
    record RewardAnalysis(
            List<List<Integer>> straightSegments,
            List<List<Integer>> tripleSegments,
            long straightReward,
            long nonStraightReward,
            long tripleReward,
            long totalReward
    ) {}

    /**
     * Result of a dice game 2.
     */
    public record DiceGame2Result(
            long guildId,
            long userId,
            List<Integer> diceRolls,
            long totalReward,
            long previousBalance,
            long newBalance,
            List<List<Integer>> straightSegments,
            List<List<Integer>> tripleSegments,
            long straightReward,
            long nonStraightReward,
            long tripleReward
    ) {
        /**
         * Formats the result as a Discord message.
         */
        public String formatMessage(String currencyIcon, String currencyName) {
            StringBuilder sb = new StringBuilder();
            sb.append("**Dice Game 2 Results**\n");
            sb.append("Rolls: ");

            for (int i = 0; i < diceRolls.size(); i++) {
                int roll = diceRolls.get(i);
                sb.append(diceEmoji(roll));
                if (i < diceRolls.size() - 1) {
                    sb.append(" ");
                }
            }

            sb.append("\n\n");

            // Show reward breakdown
            if (!straightSegments.isEmpty()) {
                sb.append(String.format("Straights: %s %,d %s\n", currencyIcon, straightReward, currencyName));
            }
            if (!tripleSegments.isEmpty()) {
                sb.append(String.format("Triples: %s %,d %s (%d group%s)\n",
                        currencyIcon, tripleReward, currencyName,
                        tripleSegments.size(), tripleSegments.size() > 1 ? "s" : ""));
            }
            if (nonStraightReward > 0) {
                sb.append(String.format("Base: %s %,d %s\n", currencyIcon, nonStraightReward, currencyName));
            }

            sb.append("\n");
            sb.append(String.format("**Total Reward:** %s %,d %s\n", currencyIcon, totalReward, currencyName));
            sb.append(String.format("**New Balance:** %s %,d %s", currencyIcon, newBalance, currencyName));

            return sb.toString();
        }

        private String diceEmoji(int value) {
            return switch (value) {
                case 1 -> ":one:";
                case 2 -> ":two:";
                case 3 -> ":three:";
                case 4 -> ":four:";
                case 5 -> ":five:";
                case 6 -> ":six:";
                default -> String.valueOf(value);
            };
        }
    }
}

File: packages/economy/src/dice/services/dice-game-2-service.ts
import {
  type Result,
  Ok,
  Err,
  DomainError,
} from '@ltdjms/shared';
import { GameRewardService } from './game-reward-service.js';
import { BalanceService } from '../../currency/services/balance-service.js';
import type { DiceGame2Config, DiceGame2Result } from '../../domain/types.js';
import {
  CurrencyTransactionSource,
  DICE_GAME_2_DICE_PER_TOKEN,
} from '../../domain/types.js';
import { type Random, DefaultRandom, rollDice } from './random.js';

/**
 * Dice Game 2 service implementation.
 * One token = 3 dice. Reward has three components:
 * 1. Straights (consecutive increasing, length >= 3): sum * straightMultiplier
 * 2. Triples (exactly 3 same values): tripleLowBonus or tripleHighBonus (sum < 10 vs >= 10)
 * 3. Remaining dice: sum * baseMultiplier
 *
 * Straights are prioritized over triples.
 * Matches Java DefaultDiceGame2Service behavior exactly.
 *
 * This service focuses on game logic only: rolling dice, analyzing rolls, and calculating rewards.
 * Config lookup and token deduction are handled by the command handler.
 */
export class DiceGame2Service {
  constructor(
    private readonly gameRewardService: GameRewardService,
    private readonly balanceService: BalanceService,
    private readonly random: Random = DefaultRandom,
  ) {}

  /**
   * Plays dice game 2. Rolls 3 dice per token,
   * analyzes straights and triples, calculates reward.
   * Config and token deduction MUST be done by the caller (handler).
   */
  async play(
    guildId: number,
    userId: string,
    tokenCount: number,
    config: DiceGame2Config,
  ): Promise<Result<DiceGame2Result, DomainError>> {
    if (tokenCount < config.minTokensPerPlay) {
      return new Err(
        DomainError.invalidInput(
          `Token count ${tokenCount} is less than minimum ${config.minTokensPerPlay}`,
        ),
      );
    }
    if (tokenCount > config.maxTokensPerPlay) {
      return new Err(
        DomainError.invalidInput(
          `Token count ${tokenCount} exceeds maximum ${config.maxTokensPerPlay}`,
        ),
      );
    }

    // Roll dice: 3 dice per token
    const diceCount = tokenCount * DICE_GAME_2_DICE_PER_TOKEN;
    const diceRolls = this.rollDice(diceCount);

    // Analyze rolls
    const analysis = this.analyzeRolls(diceRolls, config);

    // Get previous balance directly from BalanceService instead of calling
    // creditReward(0) which would trigger a full DB read with config lookup for no reward.
    const previousBalanceResult = await this.balanceService.getBalance(guildId, userId);
    const previousBalance = previousBalanceResult.isOk() ? previousBalanceResult.getValue().balance : 0;

    // Apply reward
    const rewardResult = await this.gameRewardService.creditReward(
      guildId,
      userId,
      analysis.totalReward,
      CurrencyTransactionSource.DICE_GAME_2_WIN,
    );

    if (rewardResult.isErr()) {
      return new Err(rewardResult.getError());
    }
    const newBalance = rewardResult.getValue();

    return new Ok({
      guildId,
      userId,
      diceRolls,
      totalReward: analysis.totalReward,
      previousBalance,
      newBalance,
      straightSegments: analysis.straightSegments,
      tripleSegments: analysis.tripleSegments,
      straightReward: analysis.straightReward,
      nonStraightReward: analysis.nonStraightReward,
      tripleReward: analysis.tripleReward,
    });
  }

  /**
   * Rolls the specified number of dice.
   * @internal Exposed for test use only; not part of the public API.
   */
  rollDice(count: number): number[] {
    return rollDice(count, this.random);
  }

  /**
   * Analyzes the dice rolls to identify straights, triples, and calculate rewards.
   * This logic must match Java DefaultDiceGame2Service.analyzeRolls() exactly.
   *
   * Three-pass analysis:
   * 1. Find straights (consecutive increasing sequences, length >= 3)
   * 2. Find triples among remaining dice (exactly 3 same values, non-overlapping with straights)
   * 3. Remaining dice get base multiplier
   */
  analyzeRolls(
    diceRolls: readonly number[],
    config: DiceGame2Config,
  ): {
    straightSegments: number[][];
    tripleSegments: number[][];
    straightReward: number;
    nonStraightReward: number;
    tripleReward: number;
    totalReward: number;
  } {
    const usedInStraight: boolean[] = new Array(diceRolls.length).fill(false);
    const usedInTriple: boolean[] = new Array(diceRolls.length).fill(false);

    // First pass: identify straights (consecutive increasing sequences of length >= 3)
    const straightSegments = this.findStraights(diceRolls, usedInStraight);

    // Second pass: identify triples (exactly 3 consecutive same values)
    // Must not overlap with straights
    const tripleSegments = this.findTriples(diceRolls, usedInStraight, usedInTriple);

    // Calculate rewards using configured multipliers
    const straightReward = this.calculateStraightReward(
      straightSegments,
      config.straightMultiplier,
    );
    const tripleReward = this.calculateTripleReward(
      tripleSegments,
      config.tripleLowBonus,
      config.tripleHighBonus,
    );
    const nonStraightSum = this.calculateNonStraightSum(
      diceRolls,
      usedInStraight,
      usedInTriple,
    );
    const nonStraightReward = nonStraightSum * config.baseMultiplier;

    const totalReward = straightReward + nonStraightReward + tripleReward;

    return {
      straightSegments,
      tripleSegments,
      straightReward,
      nonStraightReward,
      tripleReward,
      totalReward,
    };
  }

  /**
   * Finds all straight segments (consecutive increasing sequences of length >= 3).
   * Position-based: scans the original diceRolls array by index to find
   * consecutive increasing dice values. Matches Java
   * DefaultDiceGame2Service.findStraights() exactly.
   */
  private findStraights(
    diceRolls: readonly number[],
    usedInStraight: boolean[],
  ): number[][] {
    const straights: number[][] = [];

    // Position-based straight detection: scan the original diceRolls array
    // by index to identify contiguous segments where each die is exactly
    // one more than the previous die (consecutive increasing).
    let i = 0;
    while (i < diceRolls.length) {
      // Skip if already used in a straight (defense-in-depth for
      // non-overlapping straight segments)
      if (usedInStraight[i]) {
        i++;
        continue;
      }

      const start = i;

      // Check for consecutive increasing values in the original positions
      while (
        i + 1 < diceRolls.length &&
        diceRolls[i + 1] === diceRolls[i] + 1
      ) {
        i++;
      }

      const length = i - start + 1;

      if (length >= 3) {
        const segment: number[] = [];
        for (let j = start; j <= i; j++) {
          segment.push(diceRolls[j]);
          usedInStraight[j] = true;
        }
        straights.push(segment);
      }

      i++;
    }

    return straights;
  }

  /**
   * Finds all triple segments (exactly 3 consecutive same values).
   * A run of 4+ consecutive same values is NOT a triple.
   * Triples cannot overlap with straights.
   * Matches Java DefaultDiceGame2Service.findTriples() exactly.
   */
  private findTriples(
    diceRolls: readonly number[],
    usedInStraight: boolean[],
    usedInTriple: boolean[],
  ): number[][] {
    const triples: number[][] = [];

    let i = 0;
    while (i < diceRolls.length) {
      // Skip if already used in a straight
      if (usedInStraight[i]) {
        i++;
        continue;
      }

      const start = i;
      const value = diceRolls[i];

      // Count consecutive same values (skipping those used in straights)
      while (
        i + 1 < diceRolls.length &&
        diceRolls[i + 1] === value &&
        !usedInStraight[i + 1]
      ) {
        i++;
      }

      const length = i - start + 1;

      // Exactly 3 consecutive same values = triple
      if (length === 3) {
        const segment: number[] = [];
        for (let j = start; j <= i; j++) {
          segment.push(diceRolls[j]);
          usedInTriple[j] = true;
        }
        triples.push(segment);
      }
      // If length > 3, it''s NOT a triple - these dice will be counted as non-straight

      i++;
    }

    return triples;
  }

  /**
   * Calculates the reward for straight segments using configured multiplier.
   * Matches Java DefaultDiceGame2Service.calculateStraightReward() exactly.
   */
  private calculateStraightReward(
    straightSegments: number[][],
    straightMultiplier: number,
  ): number {
    let sum = 0;
    for (const segment of straightSegments) {
      for (const value of segment) {
        sum += value;
      }
    }
    return sum * straightMultiplier;
  }

  /**
   * Calculates the reward for triple segments using configured bonuses.
   * Each triple: if sum < 10 (values 1-3), use tripleLowBonus, otherwise tripleHighBonus.
   * Matches Java DefaultDiceGame2Service.calculateTripleReward() exactly.
   */
  private calculateTripleReward(
    tripleSegments: number[][],
    tripleLowBonus: number,
    tripleHighBonus: number,
  ): number {
    let reward = 0;
    for (const segment of tripleSegments) {
      // Each triple segment always has exactly 3 elements with the same value
      const sum = segment[0] * 3;
      if (sum < 10) {
        reward += tripleLowBonus;
      } else {
        reward += tripleHighBonus;
      }
    }
    return reward;
  }

  /**
   * Calculates the sum of dice values not used in straights or triples.
   * Matches Java DefaultDiceGame2Service.calculateNonStraightSum() exactly.
   */
  private calculateNonStraightSum(
    diceRolls: readonly number[],
    usedInStraight: boolean[],
    usedInTriple: boolean[],
  ): number {
    let sum = 0;
    for (let i = 0; i < diceRolls.length; i++) {
      if (!usedInStraight[i] && !usedInTriple[i]) {
        sum += diceRolls[i];
      }
    }
    return sum;
  }
}
